host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,430 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Invalid parameters during derivation");
+
+	/* Length of data used for derivation < key length */
+	data_len = 16;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
+		goto out;
+
+	/* Data is not multiple of 16 */
+	data_len = 18;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
+		goto out;
+
+	/* Wrong Mechanism */
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_MECHANISM_INVALID, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if operation already active");
+
+	/* Generate an AES key which can perform Encryption */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template_w_enc_der,
+			   ARRAY_SIZE(parent_template_w_enc_der),
+			   &aes_key_enc);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_EncryptInit(session, &cktest_aes_cbc_mechanism, aes_key_enc);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	data_len = 32;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key2, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_OPERATION_ACTIVE, rv))
+		goto out;
+
+	rv = C_EncryptFinal(session, NULL, NULL);
+	/* Only check that the operation is no more active */
+	if (!ADBG_EXPECT_TRUE(c, rv != CKR_BUFFER_TOO_SMALL))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if parent key C_ENCRYPT is TRUE");
+
+	data_len = 32;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key_enc, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_FUNCTION_FAILED, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if parent key C_DERIVE is FALSE");
+
+	rv = C_DestroyObject(session, aes_key1);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template_wo_derive,
+			   ARRAY_SIZE(parent_template_wo_derive),
+			   &aes_key1);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	data_len = 32;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_KEY_FUNCTION_NOT_PERMITTED, rv))
```

Comment:
`CKR_KEY_FUNCTION_NOT_PERMITTED` is not in list of expected return values for `C_DeriveKey()`.

From that list perhaps `CKR_TEMPLATE_INCONSISTENT` would be more correct.

Reply:
Good Catch. Perhaps CKR_KEY_TYPE_INCONSISTENT would be more apt.

Reply:
```
CKR_KEY_TYPE_INCONSISTENT: The specified key is not the correct type of key to use with the specified mechanism. 
This return value has a higher priority than CKR_KEY_FUNCTION_NOT_PERMITTED.
```
and
```
CKR_KEY_FUNCTION_NOT_PERMITTED: An attempt has been made to use a key for a cryptographic purpose that 
the key’s attributes are not set to allow it to do.  For example, to use a key for performing encryption, that
key MUST have its CKA_ENCRYPT attribute set to CK_TRUE (the fact that the key MUST have a CKA_ENCRYPT attribute 
implies that the key cannot be a private key).  This return value has lower priority than 
CKR_KEY_TYPE_INCONSISTENT.
```
I understand why you went for `CKR_KEY_FUNCTION_NOT_PERMITTED`. I kinda even agree with that.

But now I suppose your idea is the best matching to specification in here. Text here kinda allows "upgrade" for error. -- one could even argue that specification should be updated on the matter.

Reply:
Changed. Created a PR for this in optee_os and changed in test code.